csharp/autobuilder/Semmle.Autobuild.CSharp/DotNetRule.cs

```diff
@@ -147,11 +140,59 @@ private static BuildScript DownloadDotNet(IAutobuilder<AutobuildOptionsShared> b
                     continue;
                 }
 
-                installScript &= DownloadDotNetVersion(builder, installDir, version);
-                validGlobalJson = true;
+                return version;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Returns a script for downloading relevant versions of the
+        /// .NET Core SDK. The SDK(s) will be installed at <code>installDir</code>
+        /// (provided that the script succeeds).
+        /// </summary>
+        private static BuildScript DownloadDotNet(IAutobuilder<AutobuildOptionsShared> builder, string installDir)
+        {
+            if (!string.IsNullOrEmpty(builder.Options.DotNetVersion))
+                // Specific version supplied in configuration: always use that
+                return DownloadDotNetVersion(builder, installDir, builder.Options.DotNetVersion);
+
+            // Download versions mentioned in `global.json` files
+            // See https://docs.microsoft.com/en-us/dotnet/core/tools/global-json
+            var installScript = BuildScript.Success;
+            var foundFrameworkVersion = false;
+            var globalJsonSdkVersion = SdkVersionFromGlobalJson(builder);
+
+            if (!string.IsNullOrEmpty(globalJsonSdkVersion))
+            {
+                installScript &= DownloadDotNetVersion(builder, installDir, globalJsonSdkVersion);
+                foundFrameworkVersion = true;
+            }
+            else
+            {
+                // If there is no `global.json` to retrieve the .NET version from, we find all the
+                // TargetFramework values in the project files and install the latest versions
+                // from the corresponding channels
+                var targetFrameworkVersions = builder.ProjectsOrSolutionsToBuild
+                    .SelectMany(p => Enumerators.Singleton(p).Concat(p.IncludedProjects))
+                    .OfType<Project<CSharpAutobuildOptions>>()
+                    .Select(p => p.TargetFramework)
+                    .OfType<string>()
+                    .Where(v => v.StartsWith("net"));
+
+                foreach (var targetFrameworkVersion in targetFrameworkVersions)
+                {
+                    // For simplicity, we only accept target frameworks of the form "netX.Y"
+                    Version? version = null;
+                    if (Version.TryParse(targetFrameworkVersion[3..], out version))
```

Comment:
Consider using `out var version` and avoid declaring `version` above.
What will happen in case the target framework is `net70` and not `net7.0` (I think it will be parsed as version 70, which probably will cause problems downstream).